plugins/com.google.cloud.tools.eclipse.googleapis/src/com/google/cloud/tools/eclipse/googleapis/internal/GoogleApiFactory.java

```diff
@@ -0,0 +1,208 @@
+      return projects;
+    }
+  }
+
+  /**
+   * @return the Appengine/Apps API
+   */
+  @Override
+  public Apps newAppsApi(Credential credential) {
+    synchronized (transports) {
+      HttpTransport transport = transports.get(googleApiUrls.appEngineAdminUrl());
+      Preconditions.checkNotNull(jsonFactory, "jsonFactory is null");
+      Preconditions.checkNotNull(transport, "transport is null");
+
+      Appengine appengine =
+          new Appengine.Builder(transport, jsonFactory,
+                                credential)
+              .setApplicationName(CloudToolsInfo.USER_AGENT).build();
+      Apps apps = appengine.apps();
+      return apps;
+    }
+  }
+
+  private void buildTransports() {
+    synchronized (transports) {
+      try {
+        buildAppEngineAdminTransport();
+        buildCloudResourceManagerTransport();
+      } catch (URISyntaxException ex) {
+        logger.log(Level.SEVERE, "Could not create transport using the proxy settings", ex);
+      }
+    }
+  }
+
+  private void buildAppEngineAdminTransport() throws URISyntaxException {
+    buildTransportForUrl(googleApiUrls.appEngineAdminUrl());
+  }
+
+  private void buildCloudResourceManagerTransport() throws URISyntaxException {
+    buildTransportForUrl(googleApiUrls.cloudResourceManagerUrl());
+  }
+
+  private void buildTransportForUrl(String url) throws URISyntaxException {
+    ConnectionFactory connectionFactory =
+        new TimeoutAwareConnectionFactory(createProxy(url), DEFAULT_TIMEOUT_MS, DEFAULT_TIMEOUT_MS);
+    transports.put(url,
+                   new NetHttpTransport.Builder().setConnectionFactory(connectionFactory).build());
+  }
+
+  private Proxy createProxy(String url) throws URISyntaxException {
+    Preconditions.checkNotNull(googleApiUrls, "googleApiUrl is null");
+    Preconditions.checkArgument(!url.startsWith("http://"), "http is not supported schema");
+
+    if (proxyService == null) {
+      return Proxy.NO_PROXY;
+    }
+
+    IProxyData[] proxyDataForUri = proxyService.select(new URI(url));
+    for (final IProxyData iProxyData : proxyDataForUri) {
+      switch (iProxyData.getType()) {
+        case IProxyData.HTTPS_PROXY_TYPE:
+          return new Proxy(Type.HTTP, new InetSocketAddress(iProxyData.getHost(),
+                                                            iProxyData.getPort()));
+        case IProxyData.SOCKS_PROXY_TYPE:
+          return new Proxy(Type.SOCKS, new InetSocketAddress(iProxyData.getHost(),
+                                                             iProxyData.getPort()));
+        default:
+          logger.warning("Unsupported proxy type: " + iProxyData.getType());
+          break;
+      }
+    }
+    return Proxy.NO_PROXY;
+  }
+
+  @Reference(policy=ReferencePolicy.DYNAMIC, cardinality=ReferenceCardinality.OPTIONAL)
+  public void setProxyService(IProxyService proxyService) {
+    this.proxyService = proxyService;
+    this.proxyService.addProxyChangeListener(proxyChangeListener);
```

Comment:
I'm pretty sure OSGi ensures you won't be unbound while being bound, so this field access is safe.